src/lsm/tree.zig

```diff
@@ -227,32 +234,166 @@ pub fn TreeType(comptime Table: type) type {
+
+                context.data_block = .{
+                    .address = key_info.data_block_address,
+                    .checksum = key_info.data_block_checksum,
+                };
+
+                context.tree.grid.read_block(
+                    read_filter_block_callback,
+                    completion,
+                    key_info.filter_block_address,
+                    key_info.filter_block_checksum,
+                );
+            }
+
+            fn read_filter_block_callback(completion: *Read, filter_block: BlockPtrConst) void {
+                const context = @fieldParentPtr(LookupContext, "completion", completion);
+                assert(context.data_block != null);
+                assert(context.index_block_current < context.index_block_count);
+                assert(context.index_block_count <= config.lsm_levels);
+
+                const filter_bytes = Table.filter_block_filter_const(filter_block);
+                if (bloom_filter.may_contain(context.fingerprint, filter_bytes)) {
+                    context.tree.grid.read_block(
+                        read_data_block_callback,
+                        completion,
+                        context.data_block.?.address,
+                        context.data_block.?.checksum,
+                    );
+                } else {
+                    // The key is not present in this table, check the next level.
+                    context.advance_to_next_level();
+                }
+            }
+
+            fn read_data_block_callback(completion: *Read, data_block: BlockPtrConst) void {
+                const context = @fieldParentPtr(LookupContext, "completion", completion);
+                assert(context.data_block != null);
+                assert(context.index_block_current < context.index_block_count);
+                assert(context.index_block_count <= config.lsm_levels);
+
+                if (Table.data_block_search(data_block, context.key)) |value| {
+                    context.finish(value);
+                } else {
+                    // The key is not present in this table, check the next level.
+                    context.advance_to_next_level();
+                }
+            }
+
+            fn advance_to_next_level(context: *LookupContext) void {
+                assert(context.data_block != null);
+                assert(context.index_block_current < context.index_block_count);
+                assert(context.index_block_count <= config.lsm_levels);
+
+                context.index_block_current += 1;
+                if (context.index_block_current == context.index_block_count) {
+                    context.finish(null);
+                    return;
+                }
+                assert(context.index_block_current < context.index_block_count);
+
+                context.data_block = null;
+                context.read_index_block();
+            }
+        };
```

Comment:
Great name. Perfect control flow.

Reply:
Thanks, I was pretty happy with how this turned out :)